File: java/recursions_backtrackings/Day16/RatInMaze.java
package recursions_backtrackings.Day16;
import java.util.*;

public class RatInMaze {
    private static void solve(int i, int j, int[][] a, int n, ArrayList<String> ans, String move, int[][] vis, int[] di, int[] dj) {
        if (i == n - 1 && j == n - 1) {
            ans.add(move);
            return;
        }

        String dir = "DLRU";

        for (int ind = 0; ind < 4; ind++) {
            int nexti = i + di[ind];
            int nextj = j + dj[ind];

            if (nexti >= 0 && nextj >= 0 && nexti < n && nextj < n && vis[nexti][nextj] == 0 && a[nexti][nextj] == 1) {
                vis[nexti][nextj] = 1;

                solve(nexti, nextj, a, n, ans, move + dir.charAt(ind), vis, di, dj);

                vis[nexti][nextj] = 0;
            }
        }
    }


    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        System.out.print("Enter n: ");
        int n = sc.nextInt();

        int[][] m = new int[n][n];
        System.out.println("Enter Matrix: ");

        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                m[i][j] = sc.nextInt();
            }
        }

        ArrayList<String> ans = new ArrayList<>();

        if (m[0][0] == 0) {
            System.out.print("res: " + ans);
        }

        int[][] vis = new int[n][n];

        int[] di = {+1, 0, 0, -1};
        int[] dj = {0, -1, +1, 0};

        solve(0, 0, m, n, ans, "", vis, di, dj);

        Collections.sort(ans);

        System.out.print("Paths: " + ans);
    }
}

File: python/recursions_backtrackings/day16/ratInMaze.py
def findPath(m, n):
    ans = []

    if m[0][0] == 0:
        return ans

    visited = [[0]*n for _ in range(n)]

    di = [1, 0, 0, -1]
    dj = [0, -1, 1, 0]
    directions = "DLRU"

    def solve(i, j, path):
        if i == n-1 and j == n-1:
            ans.append(path)
            return

        for ind in range(4):
            nexti = i + di[ind]
            nextj = j + dj[ind]

            if (0 <= nexti < n and
                0 <= nextj < n and
                visited[nexti][nextj] == 0 and
                m[nexti][nextj] == 1):

                visited[nexti][nextj] = 1
                solve(nexti, nextj, path + directions[ind])
                visited[nexti][nextj] = 0

    visited[0][0] = 1
    solve(0, 0, "")

    ans.sort()
    return ans


# -------- MAIN --------
n = int(input("Enter n: "))

print("Enter matrix:")
m = []
for _ in range(n):
    row = list(map(int, input().split()))
    m.append(row)

result = findPath(m, n)

print("Paths:")
for path in result:
    print(path)